### PR TITLE
Added a new overlay helper for feedbacks

### PR DESCRIPTION
**Pull request by diegoperini**

### CODE REVIEW

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+		int containerWidth = (int) dpToPx(context, ICON_SIZE_IN_DP) + padding * 2;
+		final FrameLayout.LayoutParams containerLayoutParams = new FrameLayout.LayoutParams(
+				containerWidth,
+				FrameLayout.LayoutParams.WRAP_CONTENT,
+				OVERLAY_GRAVITY
+		);
+		containerLayoutParams.leftMargin = 0;
+		containerLayoutParams.rightMargin = 0;
+
+		for (int i = 0; i < 2; i++) {
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: Why is there a loop here? I don't see i being used at all. We just allocate a RoundedImageView and do nothing with i==0 && purpose == SCREENSHOT.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+				roundedImageView.setOnClickListener(null);
+				roundedImageView.post(new ShowHideAnimationRunnable(containerDuringSetup, roundedImageView));
+
+				container.addView(roundedImageView);
+			} else if (i == 1 && purpose == OverlayPurpose.SCREENSHOT) {
+				roundedImageView.setImageDrawable(new BitmapDrawable(context.getResources(), BitmapFactory.decodeByteArray(SCREENSHOT_ICON, 0, SCREENSHOT_ICON.length)));
+				roundedImageView.setOnClickListener(new View.OnClickListener() {
+					@Override
+					public void onClick(View v) {
+						TestFairy.showFeedbackForm();
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: We need to remove the overlay before we showFeedbackForm(). It's a one-shot button, and we can hide it before we take the screenshot.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+			}
+		}
+
+		visible = false;
+	}
+
+	static private void onResume(Activity activity) {
+		Log.d(TAG, "TestFairyFeedbackOverlay is attaching to " + activity.getClass().getSimpleName());
+
+		if (instance != null && !(activity instanceof ProvideFeedbackActivity)) {
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: If you remove the overlay prior to the push, you won't be needing the import of this class, and you won't be needing to check if activity is instanceof ProvideFeedbackActivity.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+				Paint paint = new Paint();
+				paint.setColor(ICON_BG_COLOR);
+				canvas.drawRoundRect(rect, borderRadius, borderRadius, paint);
+			}
+
+			canvas.drawRoundRect(rect, borderRadius, borderRadius, p);
+		}
+	}
+
+	static private class RoundedImageView extends ImageView {
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: Could we just have the png pre-rendered with a background color and a transparent round circle? :))

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+
+	/**
+	 * Call this before everything else to install a human friendly overlay UI
+	 * for sending feedbacks to TestFairy.
+	 *
+	 * @param context Application context or an Activity, must not be null.
+	 * @param appToken Your TestFairy app token.
+	 * @param showImmediately Set it to true of you want to show the overlay after installation completes.
+	 */
+	static public void installOverlay(Context context, String appToken, OverlayPurpose overlayPurpose, boolean showImmediately) {
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: No need for showImmediately. If somebody calls this, then they want it visible right now.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+	 * Toggles overlay visibility with animation if installed.
+	 */
+	static public void toggle() {
+		toggle(true);
+	}
+
+	/**
+	 * Toggles overlay visibility if installed.
+	 */
+	static public void toggle(boolean animated) {
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: No need for toggle.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+	 */
+	static public void hide() {
+		hide(true);
+	}
+
+	/**
+	 * Hides overlay if installed.
+	 * @param animated Whether to run an animation.
+	 */
+	static public void hide(boolean animated) {
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: No need to hide. If somebody said they want to provide feedback, they see the button. They can click on it, and then click cancel to dismiss. But there won't ever be a button on screen to remove the overlay we added.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+	static private OverlayPurpose purpose = null;
+	static private boolean visible = true;
+	static private boolean visibleBeforePause = false;
+	static private boolean beginCalled = false;
+	static private Runnable uninstall = null;
+
+	/***************** Lifecycle *****************/
+
+	static private final Application.ActivityLifecycleCallbacks lifecycle = new Application.ActivityLifecycleCallbacks() {
+		@Override
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: @Override needed in interfaces?

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+			} else {
+				Log.e(TAG, "There is no view to attach TestFairy feedback overlay");
+			}
+		}
+	}
+
+	static private void onPause(Activity activity) {
+		Log.d(TAG, "TestFairyFeedbackOverlay is detaching from " + activity.getClass().getSimpleName());
+
+		if (instance != null) {
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: No need to nullify instance. Instance will be allocated if somebody clicks a button in the app to provide feedback. From here on now, it's in memory, it barely takes up any heap. It'll be easier to handle state machine if instance is never null.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+
+		final Application a = (Application) context.getApplicationContext();
+		a.registerActivityLifecycleCallbacks(lifecycle);
+
+		visible = showImmediately;
+		visibleBeforePause = showImmediately;
+		token = appToken;
+		purpose = overlayPurpose;
+
+		uninstall = new Runnable() {
```
Comment (gmegidish) [inline app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: How come this is a variable and not a method?

**Review by gmegidish — COMMENTED** [review]
I left some review.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 1 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: The first implementation had 3 buttons. Adding multiple buttons shares the logic for layout. Once we simplifiied and reduced number of buttons to 1, this loop became obsolete. It's nice we have it in the commit history but right now it is no longer needed as you pointed out. Thanks!

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 2 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: I agree.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 3 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: Yay, less code is better ^^

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 4 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: This is what happens when a programmer designs a UI, still better than Tesla truck. :) Imma draw something with Krita, let me know how it looks in the next commit.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 5 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: Good to know.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 6 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: Less code FTW ^^

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 7 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: This means returning back from feedback form also uninstalls the overlay, right?

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 8 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: Yes.

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 9 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: Allright 👍

File: app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java (hunk 10 above)
Comment (diegoperini) [reply app/src/main/java/com/testfairy/TestFairyFeedbackOverlay.java]: Using a dispose pattern with a closure capturing a reference to `Application` simplifies avoiding a context leak. An alternative solution would be keeping a reference to Application and implementing an idempotent `uninstall()` with careful null checks. Switching from one to the other will be trivial if this class grows. Right now, I believe this is simpler.

**Pull request merged**